spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TempDirBaseInterceptor.java

```diff
@@ -0,0 +1,92 @@
+package org.spockframework.runtime.extension.builtin;
+
+import org.codehaus.groovy.runtime.ResourceGroovyMethods;
+import org.spockframework.runtime.extension.AbstractMethodInterceptor;
+import org.spockframework.runtime.extension.IMethodInvocation;
+import org.spockframework.runtime.model.SpecInfo;
+import org.spockframework.util.WrappedException;
+
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.function.BiConsumer;
+
+import static java.nio.file.FileVisitResult.CONTINUE;
+
+/**
+ * @author dqyuan
+ * @since 2.0
+ */
+public abstract class TempDirBaseInterceptor extends AbstractMethodInterceptor {
+  private static final String TEMP_DIR_PREFIX = "spock";
+
+  private final Class<?> fieldType;
+  private final BiConsumer<Object, Object> fieldSetter;
+  private final String parentDir;
+  private final boolean reserveAfterTest;
+  private Path tempPath;
+
+  TempDirBaseInterceptor(Class<?> fieldType, BiConsumer<Object, Object> fieldSetter,
+                         String parentDir, boolean reserveAfterTest) {
+    this.fieldType = fieldType;
+    this.fieldSetter = fieldSetter;
+    this.parentDir = parentDir;
+    this.reserveAfterTest = reserveAfterTest;
+  }
+
+  private Path generateTempDir(String parentDir) {
+    try {
+      if (parentDir == null || "".equals(parentDir)) {
+        return Files.createTempDirectory(TEMP_DIR_PREFIX);
+      }
+      return Files.createTempDirectory(Paths.get(parentDir), TEMP_DIR_PREFIX);
```

Comment:
Why it is not good to have path relative to the current working directory? I think it is very convenient. (Perhaps because of my inexperience >_<).

Reply:
This usually makes more problems than it provides convenience.

If this is for paths the user gives on the commandline of some tool like `doSomething --output foo/bar`, then it is great and expected to use the users working directory as base.

But for things that are "hard-coded" it is really bad.
If you for example have the files `doSomething` and `data.txt` and in `doSomething` you try to read `data.txt` with `new File("data.txt")`, then it will work in your test and as long as the one the calls `doSomething` is in the directory where those files are. But if the caller is e. g. in the parent directory and calls `directory/doSomething`, then the program will not find the file.

With Spock tests it is more the hard-coded case as the code defines the directory and the one calling does not influence the path, but influences the working directory.
Also as the value must be a compile-time constant it cannot be calculated or configured unless we would also make these closures.

But I still think that it is probably overkill to be able to configure that path per test and it would be sufficient to configure it globally in the Spock configuration file which you can extend by defining a global configuration object as documented and possible since 2.0-M4 (http://spockframework.org/spock/docs/2.0-SNAPSHOT/all_in_one.html#_configuration_objects).

Reply:
If I make `baseDir` a closure, most people will still write `@TempDir(baseDir = {Paths.get("doSomething")})`, there is no difference with compile-time constant. How to write the clousre code, so that we can avoid working directory change like you say? Thanks.

Reply:
@Vampire How to write unit test for configuration objects? I can not find example in object.

Reply:
> If I make baseDir a closure, most people will still write @TempDir(baseDir = {Paths.get("doSomething")}), there is no difference with compile-time constant. How to write the clousure code, so that we can avoid working directory change like you say? Thanks.

For example the build script could set some environment variable or system property with some absolute path and then the closure could use that variable to calculate the path where to store the temp files to. The same should be possible with the Spock configuration file I think, as it is a Groovy script.

> @Vampire How to write unit test for configuration objects? I can not find example in object.

This is one of the occasions where you actually need the `EmbeddedSpecRunner`.
Have a look at `UnrolledFeatureMethods` or `UprolledFeatureMethods` specifications.
There I tested the configuration script options for uprolling and unrolling.